lib/ancestry/class_methods.rb

```diff
@@ -69,28 +64,19 @@ def arrange_serializable options={}, nodes=nil, &block
 
     # Pseudo-preordered array of nodes.  Children will always follow parents,
     # for ordering nodes within a rank provide block, eg. Node.sort_by_ancestry(Node.all) {|a, b| a.rank <=> b.rank}.
+    # TODO: Fix case when parents are missing and a sort is specified
     def sort_by_ancestry(nodes, &block)
-      arranged = nodes if nodes.is_a?(Hash)
-
-      unless arranged
-        presorted_nodes = nodes.sort do |a, b|
-          a_cestry, b_cestry = a.ancestry || '0', b.ancestry || '0'
-
-          if block_given? && a_cestry == b_cestry
-            yield a, b
-          else
-            a_cestry <=> b_cestry
-          end
+      if nodes.is_a?(Hash) || block_given?
+        nodes = arrange_nodes(nodes) unless nodes.is_a?(Hash)
+
+        nodes = nodes.sort { |(a, a_children), (b, b_children)| yield(a, b) } if block_given?
+        nodes.inject([]) do |sorted_nodes, (node, children)|
+          sorted_nodes << node
+          sorted_nodes += sort_by_ancestry(children, &block) unless children.blank?
+          sorted_nodes
```

Comment:
Since you are always appending in some form, can you use flat_map?

Reply:
Oh nvm, this was previous code... need to try ?w=1

Reply:
I was thinking of passing `sorted_nodes` as a method parameter.
I had tried to write as a `flat_map` but my brain just couldn't wrap it.
Suggestions welcome here (even if it is "unchanged")